### PR TITLE
E2E test for irqwatch tool

### PR DESCRIPTION
**Pull request by Tal-or**
The test will generate a significant amount of random changes in the interrupt counters`
and compare the delta from the irqwatch tool with the expected delta that the test created

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

### CODE REVIEW

File: test/e2e/knit_irqwatch_test.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+	return irqS.Counters, nil
+}
+
+func stats2JSON(s irqs.Stats) ([]byte, error) {
+	cpus := cpuset.CPUSet{}
+	cnt := struct {
+		Counters irqs.Stats
+	}{
+		Counters: cm.CountersForCPUs(cpus, s),
```
Comment (Tal-or) [inline test/e2e/knit_irqwatch_test.go:221]: @fromanirh  ```CountersForCPUs``` was taken from [here](https://github.com/fromanirh/debug-tools/blob/irqwatch-cmd/pkg/knit/cmd/irqwatch.go#L226)
Please replace it with a capital 'C'

File: test/e2e/knit_irqwatch_test.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+			}
+			fmt.Fprintf(g.GinkgoWriter, "running: %v\n", cmdline)
+
+			cmd := exec.Command(cmdline[0], cmdline[1:]...)
+			cmd.Stderr = g.GinkgoWriter
+
+			c := make(chan cmdOutput)
+			go func(c chan cmdOutput) {
+				out, err := cmd.Output()
+				c <- cmdOutput{out: out,
```
Comment (fromanirh) [inline test/e2e/knit_irqwatch_test.go:50]: add a newline so it reads like
```
c <- cmdOutput{
    out: out,
    err: err.
}

File: test/e2e/knit_irqwatch_test.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+				"-T", "1",
+				"-J",
+				"irqwatch",
+			}
+			fmt.Fprintf(g.GinkgoWriter, "running: %v\n", cmdline)
+
+			cmd := exec.Command(cmdline[0], cmdline[1:]...)
+			cmd.Stderr = g.GinkgoWriter
+
+			c := make(chan cmdOutput)
```
Comment (fromanirh) [inline test/e2e/knit_irqwatch_test.go:47]: isn't `cmd.Output()` running on a different goroutine?

File: test/e2e/knit_irqwatch_test.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+		snapshotRoot = snapshotBeforeEach(fixtureName, "sysinfo.tgz")
+	})
+
+	g.AfterEach(func() {
+		snapshotAfterEach(snapshotRoot)
+	})
+})
+
+func modifyInterruptsFile(procFs string) (irqs.Stats, error) {
+	mockLog := log.New(ioutil.Discard, "", 0)
```
Comment (fromanirh) [inline test/e2e/knit_irqwatch_test.go:82]: I'd call this `fakeLog`

File: test/e2e/knit_irqwatch_test.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		delta[cpuid] = make(irqs.Counter)
+		delta[cpuid][k] = stats[cpuid][k] - initialStats[cpuid][k]
+	}
+
+	// add the random new stat values to the interruts file
+	return delta, reWriteInterrupts(procFs, stats)
+}
+
+func reWriteInterrupts(procFs string, stats irqs.Stats) error {
+	tmpInterrruptsFile := filepath.Join(procFs, "interrupts_temp")
```
Comment (fromanirh) [inline test/e2e/knit_irqwatch_test.go:117]: I'd use the utilities in `io/ioutil`

File: test/e2e/knit_irqwatch_test.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	os.Rename(tmpInterrruptsFile, interrruptsFile)
+	if err != nil {
+		g.Fail(fmt.Sprintf("fail to rename file %q to %q", tmpInterrruptsFile, interrruptsFile))
+		return err
+	}
+
+	return nil
+}
+
+func tab(f *os.File) {
```
Comment (fromanirh) [inline test/e2e/knit_irqwatch_test.go:171]: please inline in the calling sites

File: test/e2e/knit_irqwatch_test.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	f.WriteString("\t")
+}
+
+func newLine(f *os.File) {
```
Comment (fromanirh) [inline test/e2e/knit_irqwatch_test.go:175]: please inline in the calling sites

File: test/e2e/knit_irqwatch_test.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	f.WriteString("\n")
+}
+
+func selRandKey(c irqs.Counter) string {
+	//fmt.Println(len(c))
```
Comment (fromanirh) [inline test/e2e/knit_irqwatch_test.go:180]: please remove commented code

File: test/e2e/knit_irqwatch_test.go [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+
+func tab(f *os.File) {
+	f.WriteString("\t")
+}
+
+func newLine(f *os.File) {
+	f.WriteString("\n")
+}
+
+func selRandKey(c irqs.Counter) string {
```
Comment (fromanirh) [inline test/e2e/knit_irqwatch_test.go:179]: I'd use something along the lines of https://stackoverflow.com/questions/33994677/pick-a-random-value-from-a-go-slice
1. make a slice of keys
2. pick a random key using `rand.Intn`

**Review by fromanirh — COMMENTED** [review]
few inline comments regarding the implementation

File: test/e2e/knit_irqwatch_test.go (hunk 3 above)
Comment (Tal-or) [reply test/e2e/knit_irqwatch_test.go:47]: It does, but I provide the channel ```cmdOutput``` as an [argument](https://github.com/openshift-kni/debug-tools/pull/26/files#diff-105a28bc7bb59f3ba60810661320b99cfa11962075cc2117b2f504af0cf736ebR53) to the function that goroutine runs

File: test/e2e/knit_irqwatch_test.go (hunk 5 above)
Comment (Tal-or) [reply test/e2e/knit_irqwatch_test.go:117]: not sure I follow, what alternative has to ```io/ioutil``` to offer?

Comment (fromanirh) [reply test/e2e/knit_irqwatch_test.go:117]: https://golang.org/pkg/io/ioutil/#TempFile + `os.Rename` once done. This is pretty much the standard pattern

File: test/e2e/knit_irqwatch_test.go (hunk 6 above)
Comment (fromanirh) [reply test/e2e/knit_irqwatch_test.go:171]: also, in general better to abstract and use `w io.Writer` as argument - but again, let's inline this code anyway.

**Comment by fromanirh** [comment]
/lgtm
thanks for the changes, if CI likes it we can move forward

**Pull request merged**